### DISCUSSION TITLE
FastAPI: Router APIs is not working in cpanel

### DISCUSSION THREAD

**Discussion by hamzakt21** [fastapi/fastapi#6392] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# passenger_wsgi.py code
from a2wsgi import ASGIMiddleware
from app.main import app

application = ASGIMiddleware(app)


# main.py code
from fastapi import FastAPI
from .routers import authentication
import uvicorn

app = FastAPI(debug=True)

app.include_router(authentication.router)

@app.get("/", tags=['Welcome'])
def index():
    return {"detail": "Welcome to APIs."}
```

### Description

Hello Folks,

I have developed one application in FastAPI and it is working fine on my local machine. After hosting that application in Cpanel, main.py API is working but router APIs are not working. Not working means when I request URL it tell me "Not found".

I have followed the file structure that is given in the FastAPI documentation.

Please let me know how can I resolve this issue.

Thanks.

### Operating System

Linux

### FastAPI Version

0.75.1

### Python Version

3.7

**Comment by yigitesen** [comment]
Hello,

Were you able to find any solution? i am having the same problem.

**Comment by hamzakt21** [comment]
Hello yigitesen, Not yet. For the solution, I have moved to the Linux VPS, and using gunicorn I have deployed the application.

**Comment by yigitesen** [comment]
> Hello yigitesen, Not yet. For the solution, I have moved to the Linux VPS, and using gunicorn I have deployed the application.

Thank you for response. I will try that.

**Comment by yigitesen** [comment, marked as answer]
Hello again hamzakt21,
I got the solution. In your case routers folder should be in the src folder and it should contain __init__.py after theese APIRouter works for my case.

**Comment by hamzakt21** [comment]
Okay. Thank you so much @yigitesen. It is really helpful.

**Comment by Tusar0003** [comment]
hello @yigitesen 
can you help me??? i am facing the same issue cpanel. it keeps on loading and then showing request timeout.

can you be more specific what actually needs to be done??

**Comment by mdkberry** [comment]
I second this request to know more about how to get this working. The answer provided is very vague and different to what I am seeing in python app on CPANEL on a linux web server host.

**Comment by naveed-patel** [comment]
Create your application (https://fastapi.tiangolo.com/tutorial/bigger-applications/) inside the app folder and passenger_wsgi.py as below. Also touch __init__.py in the root folder next to passenger_wsgi.py


```
from a2wsgi import ASGIMiddleware
from app.main import app
application = ASGIMiddleware(app)
```


The directory listing would be as 


```
.:
app  __init__.py  passenger_wsgi.py 

./app:
dependencies.py  __init__.py  internal  main.py  routers

./app/internal:
admin.py  __init__.py

./app/routers:
__init__.py  items.py  users.py
```

**Comment by Tusar0003** [comment]
I don't know why it's still not working. It just keeps on loading.......if you could me tell what's wrong with my code or directory settings, it would be really helpful

![Screenshot 2022-09-02 163249](https://user-images.githubusercontent.com/18624915/188121562-65951be7-6287-41ff-b415-22870571d888.png)
![Screenshot 2022-09-02 163416](https://user-images.githubusercontent.com/18624915/188121565-a26e3de0-0fa4-4a91-9324-e2038fed11fd.png)

![Screenshot 2022-09-02 163047](https://user-images.githubusercontent.com/18624915/188121554-8ee7e88d-7f88-494a-9460-7bb3dcba6dd9.png)
![Screenshot 2022-09-02 163107](https://user-images.githubusercontent.com/18624915/188121560-7e4cfd64-a28b-4170-a3f5-a4a2a52280f1.png)

**Comment by naveed-patel** [comment]
I don't see any error with your code. I think the python passenger application isn't set up correctly and it never tries to start it. If it would have, you would have seen __pycache__ in the root folder. Have you created the python app in cpanel. If yes, best to check how to get this app set up with the hosting provider.

**Comment by sysalong** [comment]
can you help me??? i am facing the same issue cpanel. it keeps on loading and then showing request timeout.

can you be more specific what actually needs to be done??

**Comment by mdkberry** [comment]
I am back looking at this again, so will post some things I have found out in case it helps others. Mine still is not working properly yet, but my layout is slightly different. Though it times out when I try to access the site, interestingly if I enable the debugging section in passenger_wsgi.py then that works to show my python version and a message saying "It works!" but as soon as I comment out the debugging I am back to timeout in the browser. I'll add more if I ever solve it.

this is my folder structure at top level for my api website:
![top level folder](https://user-images.githubusercontent.com/810955/223898943-5e97b4c1-c5ff-480a-9ca7-2ffa2f840cfa.png)

this is the python app in cpanel:
![python app setup cpanel](https://user-images.githubusercontent.com/810955/223932578-98f55aff-27c3-4e5e-a330-ea2ad141f9d1.png)


my passenger_wsgi.py:

```python
import os
import sys
from a2wsgi import ASGIMiddleware
from main import app  # Import your FastAPI app.


application = ASGIMiddleware(app)

'''
# debugging section
async def log_middleware(scope, receive, send):
    async def _receive():
        msg = await receive()
        print(msg)
        return msg

    async def _send(msg):
        print(msg)
        await send(msg)

    return await app(scope, _receive, _send)

# end of debugging section


sys.path.insert(0, os.path.dirname(__file__))


def application(environ, start_response):
    start_response('200 OK', [('Content-Type', 'text/plain')])
    message = 'It works!\n'
    version = 'Python %s\n' % sys.version.split()[0]
    response = '\n'.join([message, version])
    return [response.encode()]
'''
```

and my main.py:

```python
from fastapi import FastAPI
import uvicorn

app = FastAPI()

@app.get("/")
async def root():    
    return "Hello World!"
```

**Comment by mdkberry** [comment]
some additional things to note. In the above CPanel entry for the python app, you can take the URL code at the top and use it in CPanel terminal to access your python environment (I have had to do this because my SSH has access issues I havent yet resolved). Once there check `pip install uvicorn` and `pip install fastapi` is done.

here you can also run `ps aux | grep uvicorn` to see if your process is running. mine wasnt.

One problem I ran into is that uvicorn wont run on port 80, but it will run on port 8000 or port 8080 and I believe this is restricted by my web host provider so I am negotiating with them to open one of the latter ports.

you can test it with `uvicorn main:app --host 0.0.0.0 --port 8000` and change the port to see what results you get. Also then check your web site for open ports to confirm access or not.

that is as far as I got so far. still not working, but closer to getting it there.

**Reply by mdkberry** [reply]
It looks like I wont be continuing with this approach anyway given this response from my web host provider, so I either need to host this myself, which means approaching it differently, or delving into other aspects which I dont have time right now to do:

*"Opening these ports will not be possible, and I'll provide an example of why. You're requesting port 8080 to be open for you. It would mean that no other user could use that port on the server you're on. That would also mean we'd now keep track of the various ports every user uses.* 

*Our Python/Ruby/Nodejs application systems allow you to run an application even if it expects to bind to port 80 or another port. The implementation acts similarly to Phusion Passenger, which will attach to your application port and serve requests from our web server to yours. When you create the application, you'll find in the document root you suggested that the .htaccess file has added Passenger rules to allow this all to interact."*

**Comment by DCypher777** [comment]
Hey guys. I just saw this thread today. I tried it this way and it worked. 
# main.py
from fastapi import FastAPI
import uvicorn

app = FastAPI()

@app.get("/")
async def root():    
    return "Hello World!"

Note that ASGI Middleware is required for Asynchronous route calls 
# passenger_wsgi.py code
from a2wsgi import ASGIMiddleware
from main import app

application = ASGIMiddleware(app)

**Reply by DCypher777** [reply]
Ps. for CPanel: 
Application Startup file is: main.py 
Application Entry Point is: passenger_wsgi.py

**Reply by one-numan** [reply]
I followed yours step and Deployed Code in Cpanel this but still Showing "500 Internal Server Error "

- How to Solve this Issue and What are reasons or corner case that cause Raise error

**Comment by one-numan** [comment]
I followed https://github.com/tiangolo/fastapi/discussions/6392#discussioncomment-5846032 and Deployed Code in Cpanel this but still Showing "500 Internal Server Error "

- How to Solve this Issue and What are reasons or corner case that cause Raise error

**Reply by Cavein254** [reply]
In my case the deployment works fine but the problem is when I reload a page, the response takes forever.
`This request takes too long to process, it is timed out by the server`

**Reply by SE-Spark** [reply]
> In my case the deployment works fine but the problem is when I reload a page, the response takes forever. `This request takes too long to process, it is timed out by the server`

did you find any work around?

**Comment by tranngocminhhieu** [comment]
Hello everyone. I found a solution in [How to deploy FastAPI on cPanel](https://lucidgen.com/en/how-to-deploy-fastapi-on-cpanel/). It works for me.
